analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyLoggingCalls.cs

```diff
@@ -0,0 +1,217 @@
+    public int WarningThreshold { get; set; } = DefaultTresholds.Warning;
+
+    protected override void Initialize(SonarParametrizedAnalysisContext context) =>
+        context.RegisterCompilationStartAction(cc =>
+        {
+            if (cc.Compilation.ReferencesAny(KnownAssembly.CastleCore, KnownAssembly.Log4Net, KnownAssembly.MicrosoftExtensionsLoggingAbstractions, KnownAssembly.Serilog, KnownAssembly.NLog))
+            {
+                cc.RegisterNodeAction(Process, SyntaxKind.Block, SyntaxKind.CompilationUnit);
+            }
+        });
+
+    private void Process(SonarSyntaxNodeReportingContext context)
+    {
+        var node = context.Node;
+        if (IsBlockNodeSupported(node))
+        {
+            var logCallCollector = new LoggingCallCollector(context.SemanticModel, node);
+            logCallCollector.Visit(node);
+            foreach (var loggingCategory in logCallCollector.LoggingInvocations)
+            {
+                var treshold = Treshold(loggingCategory.Key);
+                var invocations = loggingCategory.Value;
+                if (invocations.Count > treshold)
+                {
+                    var primaryLocation = invocations[0].GetLocation();
+                    var secondaryLocations = invocations.Skip(1).Select(x => x.GetLocation());
+                    context.ReportIssue(Diagnostic.Create(Rule, primaryLocation, secondaryLocations, loggingCategory.Key, invocations.Count, treshold));
+                }
+            }
+        }
+    }
+
+    private int Treshold(string category)
+    {
+        var threshold = category switch
+        {
+            CategoryNames.Debug => DebugThreshold,
+            CategoryNames.Error => ErrorThreshold,
+            CategoryNames.Warning => WarningThreshold,
+            _ => InformationThreshold
+        };
+        return Math.Max(threshold, 0);
+    }
+
+    private static bool IsBlockNodeSupported(SyntaxNode node) =>
+        node.Kind() == SyntaxKind.Block
+        || (node.Kind() == SyntaxKind.CompilationUnit && node.ChildNodes().OfType<GlobalStatementSyntax>().Any());  // for top-level statements
+
+    private sealed class LoggingCallCollector : SafeCSharpSyntaxWalker
+    {
+        private readonly SemanticModel semanticModel;
+        private SyntaxNode currentBlock;
+
+        public Dictionary<string, List<InvocationExpressionSyntax>> LoggingInvocations { get; } = [];
+
+        public LoggingCallCollector(SemanticModel semanticModel, SyntaxNode currentBlock)
+        {
+            this.semanticModel = semanticModel;
+            this.currentBlock = currentBlock;
+        }
+
+        public override void VisitBlock(BlockSyntax node)
+        {
+            if (currentBlock == node) // Don't visit nested blocks, that will be done by additional LoggingCallCollector instances.
+            {
+                currentBlock = node;
```

Comment:
This assignment confuses me.
Isn't it redundant?

Reply:
It will prevent the Syntax Walker from processing nested blocks.

```cs
void Method(ILogger logger, bool condition)
{
    logger.Error("Error 1");
    if (condition)
    {
       logger.Error("Error 2");
    }
}
```

In this method `RegisterNodeAction` will first process the outer block with a Syntax Walker, then later the inner block with another instance of the Syntax Walker. I don't want the first Syntax Walker to go into the nested block because then it would count 2 Error logging methods even though they are in separate blocks.
So when the Syntax Walker visits a `Block` it always checks if it's the one it should visit (it's passed in the Walker's constructor), and if it's not then it won't visit that branch of the syntax tree. The additional catch is that this can either be a `Block` or a `CompilationUnit` (root of the syntax tree when traversing a code with top-level statements).

Reply:
I understand why you need to do the check, but I cannot think of a reason to do this:

```csharp
if (currentBlock == node)
{
    currentBlock = node; // <- this seems useless, since you just asserted that they are equal.
    base.VisitBlock(node);
}
```

Reply:
I see now. You're right, I accidentally left it there from a previous implementation. Now I removed it.